cheatsheets/Protect_FileUpload_Against_Malicious_File.md

```diff
@@ -1,597 +1,138 @@
+    if "file.php".endswith(bad):
+        // raise exception
 ```
 
-# Sources of the prototype
+### User Permissions
+
+Before any file upload service is accessed, proper validation should occur on two levels for the user uploading a file:
+
+- Authentication level
+  - The user should be a registered user, or an identifiable user, in order to set restrictions and limitations for their upload capabilities
+- Authorization level
+  - Make sure that the user's accesses allow them to upload a file, or multiple files for administrative cases.
+
+### Content-Type Validation
+
+Other than defining the extension of the uploaded file, its MIME-type can be checked for a quick protection against simple file upload attacks.
+
+This can be done preferrably in a whitelist approach; otherwise, this can be done in a blacklist approach.
+
+> This should not be used on its own, as bypassing it is pretty common and easy.
+
+### File Name Sanitization
+
+Multiple threats relate to the file-name, thus, it is essential to ensure that the file-name is stored in a secure fashion.
+
+This can be done by creating a **random string** as a file-name. If the file-name is required by the business needs, proper input validation should be done for client-side (_e.g._ active content) and back-end side (_e.g._ special files overwrite or creation) attack vectors. File-name length limits should be taken into consideration based on the system storing the files, as each system has its own file name length limit.
+
+### File Storage Location
+
+All file uploaded will be stored in a location on the server. This location must be chosen based on security and business requirements. The following points are set by security priority, and are inclusive:
+
+1. Store the files on a **different host**, which allow for complete segragation of duties between the application serving the user, and the host handling file uploads and their storage.
+2. Store the files in a **private directory**, where only administrative access is allowed.
+3. Store the files on a **public directory**, and set them in write permissions only.
+
+> If read access is required, setting proper controls is a must (_e.g._ internal IP, authorized user, etc.)
+
+### System Permissions
+
+Files should be stored in a way that ensures:
+
+- Allowed system users are the only ones capable of reading the files
+- Required modes only are set for the file
+  - If execution is required, scanning the file before running it is required as a security best practice, to ensure that no macros or hidden scripts are available.
+
+### Upload and Download Limits
+
+The application should set proper size limits for the upload service in order to protect the file storage capacity. The file size limit should be considered after file decompression is conducted, and by using secure methods to calculate zip files size.
+
+It should set proper request limits as well for the download service if available to protect the server from DoS attacks.
```

Comment:
I'd be tempted to remove the stuff about downloading - this is a whole separate area with **lots** of potential issues. Perhaps a separate CS?

Reply:
What would we be tackling specifically in there? To see if it requires a whole CS, or a section.

Reply:
Planning to rename this CS too, once we're done. Can be something to cover both.

Reply:
It's often an area that I see very bad vulns around when doing pentests. Areas such as:

* Directory traversal/arbitrary download
* Downloads from other servers (SSRF, `file://` URIs, etc)
* Filename tampering (e.g, having files downloaded with dangerous extensions such as .bat, CRLF injection, etc)
* Access control issues on a file download handler
* No access control for files in the webroot

Perhaps not a CS in its own right, but certainly a decent size section if it's in another one.

Maybe this could become a combined "file upload and download" or something?

Reply:
For now, I will keep the small note. It can be tackled in another PR. As I said previously, this CS can handle both. We just need to change the name and make it firm in the ground.

Reply:
Should we create an issue for it?